### PR TITLE
Return shared_pointer in Python WorkspaceGroup.getItem()

### PR DESCRIPTION
**Pull request by DavidFair**
**Description of work.**

See commit message for in-depth explanation. This has resolved the regression seen in #29939 

**To test:**
- Ensure new unit test passes

*There is no associated issue, this was noticed whilst working on something else*

This does not require release notes because it's obscure enough few people will have run into it outside the development team. Especially as it requires people to only use workspace pointers rather than the ADS

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: Framework/PythonInterface/test/python/mantid/api/WorkspaceGroupTest.py [hunk 1]
```diff
@@ -34,10 +35,30 @@ def test_WorkspaceGroup_can_be_instantiated(self):
...
+        def __init__(self):
+            self.group = WorkspaceGroup()
+
+        def __enter__(self):
+            return self.group
+
+        def __exit__(self, exc_type, exc_val, exc_tb):
+            del self.group  # Make our intention explicit for future readers
+
+    def test_that_get_item_scope_works_not_in_ads(self):
```
Comment (martyngigg) [inline Framework/PythonInterface/test/python/mantid/api/WorkspaceGroupTest.py:52]: Nice test!

File: Framework/PythonInterface/mantid/api/src/Exports/WorkspaceGroup.cpp [hunk 2]
```diff
@@ -90,7 +89,6 @@ void export_WorkspaceGroup() {
       .def("remove", &WorkspaceGroup::remove,
            (arg("self"), arg("workspace_name")), "Remove a name from the group")
       .def("getItem", getItem, (arg("self"), arg("index")),
-           return_value_policy<Policies::ToWeakPtr>(),
```
Comment (martyngigg) [inline Framework/PythonInterface/mantid/api/src/Exports/WorkspaceGroup.cpp:93]: Thanks for finding this whole in the setup.

I see what you're saying regarding the fixed type requirement in the policy as it needs to return the correct thing from `get_pytype()`. I think in this case we could use the local `getItem` intermediary that is already defined above. If we alter that to return a `boost::python::object` directly then we don't need the policy and we can decide in that function how to do the conversion. What do you think?

My main concern is interface code that holds onto one of these references that then means deleting something in the workspaces view will leave a workspacin memory with no way to delete it.

Comment (DavidFair) [reply Framework/PythonInterface/mantid/api/src/Exports/WorkspaceGroup.cpp:93]: Actually that's a really elegant way to fix it, thanks.
Looking back I think I was so busy trying to write a generic wrapper to decay to weak pointer. Coming back to it, as of this PR, this is the only place we decayed to a weak pointer. So taking the agile approach of not trying to anticipate future needs made it much easier.

**Review by martyngigg — APPROVED** [review]
Looks good now. Thanks for making those changes.

**Pull request merged**